Config controller dev

csookarr opened the pull request.

csookarr committed:
Add rmd-pod-ds.yaml

csookarr committed:
Add rmd-node-agent-ds.yaml

csookarr committed:
RMD Config controller up and running (WIP)

csookarr committed:
RMD Config controller runs without error and README updated

csookarr committed:
Use IP Adresses instead of names to find pods

nolancon commented on build/Dockerfile:
```diff
@@ -25,8 +25,8 @@ ENV OPERATOR=/usr/local/bin/intel-rmd-operator \
-COPY build/manifests/rmd-node-agent.yaml /rmd-manifests/rmd-node-agent.yaml
-COPY build/manifests/rmd-pod.yaml /rmd-manifests/rmd-pod.yaml
+COPY build/manifests/rmd-node-agent-ds.yaml /rmd-manifests/rmd-node-agent-ds.yaml
+COPY build/manifests/rmd-pod-ds.yaml /rmd-manifests/rmd-pod-ds.yaml
```
Simply `rmd-ds.yaml` is better

nolancon commented on pkg/controller/add_node.go:
```diff
@@ -1,10 +1,11 @@
 package controller
 
+/*
 import (
 	"github.com/intel/rmd-operator/pkg/controller/node"
 )
 
 func init() {
 	// AddToManagerFuncs is a list of functions to create controllers and add them to a manager.
 	AddToManagerFuncs = append(AddToManagerFuncs, node.Add)
-}
+}*/
```
We will need to remove/tidy up files like this related to the old node controller. For now it's ok though... just making a note for future reference.

nolancon commented on pkg/controller/rmdconfig/rmdconfig_controller.go:
```diff
@@ -31,8 +36,8 @@ const (
 
-var rmdPodPath = "/rmd-manifests/rmd-pod.yaml"
-var nodeAgentPath = "/rmd-manifests/rmd-node-agent.yaml"
+var rmdDaemonSetPath = "/rmd-manifests/rmd-pod-ds.yaml"
```
As above - `rmd-ds.yaml`

nolancon commented on pkg/controller/rmdconfig/rmdconfig_controller.go:
```diff
@@ -138,19 +146,209 @@ func (r *ReconcileRmdConfig) Reconcile(request reconcile.Request) (reconcile.Res
+		nodeName := string(node.GetObjectMeta().GetName())
+		rmdNamespacedName := types.NamespacedName{
+			Namespace: defaultNamespace,
+			Name:      "rmd-daemon-set",
```
simply `"rmd"` for `Name:` and should be defined at top of file as a `const`

nolancon commented on pkg/controller/rmdconfig/rmdconfig_controller.go:
```diff
@@ -138,19 +146,209 @@ func (r *ReconcileRmdConfig) Reconcile(request reconcile.Request) (reconcile.Res
-
+		nodeAgentNamespacedName := types.NamespacedName{
+			Namespace: defaultNamespace,
+			Name:      "node-agent-daemon-set",
```
simply `"rmd-node-agent"` for `Name:` and should be defined at top of file as a const

nolancon commented on pkg/controller/rmdconfig/rmdconfig_controller.go:
```diff
@@ -138,19 +146,209 @@ func (r *ReconcileRmdConfig) Reconcile(request reconcile.Request) (reconcile.Res
+			Namespace: defaultNamespace,
+			Name:      "rmd-daemon-set",
+		}
+		err = r.createDaemonSetIfNotPresent(&node, rmdNamespacedName, rmdDaemonSetPath)
```
this function should accept the RmdConfig object that is being reconciled instead of `node` - this will be clearer below.

nolancon commented on pkg/controller/rmdconfig/rmdconfig_controller.go:
```diff
@@ -138,19 +146,209 @@ func (r *ReconcileRmdConfig) Reconcile(request reconcile.Request) (reconcile.Res
+			Namespace: defaultNamespace,
+			Name:      "node-agent-daemon-set",
+		}
+		err = r.createDaemonSetIfNotPresent(&node, nodeAgentNamespacedName, nodeAgentDaemonSetPath)
```
same as above comment on `r.createDaemonSetIfNotPresent(...`

nolancon commented on pkg/controller/rmdconfig/rmdconfig_controller.go:
```diff
@@ -138,19 +146,209 @@ func (r *ReconcileRmdConfig) Reconcile(request reconcile.Request) (reconcile.Res
 	return reconcile.Result{}, nil
 }
+
+func (r *ReconcileRmdConfig) createDaemonSetIfNotPresent(node *corev1.Node, namespacedName types.NamespacedName, path string) error {
```
should accept `rmdConfig *intelv1alpha1.RmdConfig` instead of `node *corev1.Node`

nolancon commented on pkg/controller/rmdconfig/rmdconfig_controller.go:
```diff
@@ -138,19 +146,209 @@ func (r *ReconcileRmdConfig) Reconcile(request reconcile.Request) (reconcile.Res
+				logger.Error(err, "Failed to build daemonSet from manifest")
+				return err
+			}
+			if err := controllerutil.SetControllerReference(node, daemonSet, r.scheme); err != nil {
```
Now we can replace `node` with `rmdConfig`
This makes rmdConfig the owner of the daemonset - this means if rmdConfig is deleted, the daemonset is also garbage collected.

nolancon commented on pkg/controller/rmdconfig/rmdconfig_controller.go:
```diff
@@ -138,19 +146,209 @@ func (r *ReconcileRmdConfig) Reconcile(request reconcile.Request) (reconcile.Res
+		rmdNodeStateNamespacedName := types.NamespacedName{
+			Namespace: defaultNamespace,
+			Name:      rmdNodeStateName,
+		}
```
this can be moved into `r.createNodeStateIfNotPresent(..` function. Will explain below.

nolancon commented on pkg/controller/rmdconfig/rmdconfig_controller.go:
```diff
@@ -138,19 +146,209 @@ func (r *ReconcileRmdConfig) Reconcile(request reconcile.Request) (reconcile.Res
+	return nil
+}
+
+func (r *ReconcileRmdConfig) createNodeStateIfNotPresent(node *corev1.Node, namespacedName types.NamespacedName) error {
```
this function should accept the following parameters:
`(node *corev1.Node, rmdConfig *intelv1alpha1.RmdConfig`

We can then remove the `nodeName := string(node.GetObjectMeta().GetName())` call at the top of the loop and place it in here to construct the `namespacedName` locally inside this function.

nolancon commented on pkg/controller/rmdconfig/rmdconfig_controller.go:
```diff
@@ -138,19 +146,209 @@ func (r *ReconcileRmdConfig) Reconcile(request reconcile.Request) (reconcile.Res
+			}
+			workloads := make(map[string]intelv1alpha1.WorkloadMap)
+			rmdNodeState.Status.Workloads = workloads
+			if err := controllerutil.SetControllerReference(node, rmdNodeState, r.scheme); err != nil {
```
Also replace `node` with `rmdConfig` here.

nolancon commented on pkg/controller/rmdconfig/rmdconfig_controller.go:
```diff
@@ -138,19 +146,209 @@ func (r *ReconcileRmdConfig) Reconcile(request reconcile.Request) (reconcile.Res
+			Namespace: defaultNamespace,
+			Name:      rmdNodeStateName,
+		}
+		err = r.createNodeStateIfNotPresent(&node, rmdNodeStateNamespacedName)
```
this function will accept parameters `node` and `rmdConfig` - explained below at function.

nolancon commented on pkg/controller/rmdconfig/rmdconfig_controller.go:
```diff
@@ -138,19 +146,209 @@ func (r *ReconcileRmdConfig) Reconcile(request reconcile.Request) (reconcile.Res
+		logger.Info("Failed to list Pods")
+		return err
+	}
+	podName := ""
+	for _, pod := range pods.Items {
+		for _, address := range rmdNode.Status.Addresses {
+			if address.Address == pod.Status.HostIP {
+				podName = pod.GetObjectMeta().GetName()
+				break
+			}
+		}
+	}
```
Should make this its own function and `return podName` once found instead of `break` to save iterations.

nolancon commented on pkg/controller/rmdnodestate/rmdnodestate_controller.go:
```diff
@@ -112,7 +112,35 @@ func (r *ReconcileRmdNodeState) Reconcile(request reconcile.Request) (reconcile.
 
-	rmdPodName := fmt.Sprintf("%s%s", rmdPodNameConst, rmdNodeState.Spec.Node)
+	rmdNode := &corev1.Node{}
+	nodeName := strings.ReplaceAll(request.Name, "rmd-node-state-", "")
+	nodeNamespacedName := types.NamespacedName{
+		Namespace: "",
+		Name:      nodeName,
+	}
+	err = r.client.Get(context.TODO(), nodeNamespacedName, rmdNode)
```
I think this can be condensed into:
`err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: "", Name: rmdNodeState.Spec.Node}, rmdNode)`

nolancon commented on pkg/controller/rmdnodestate/rmdnodestate_controller.go:
```diff
@@ -112,7 +112,35 @@ func (r *ReconcileRmdNodeState) Reconcile(request reconcile.Request) (reconcile.
+		return reconcile.Result{}, err
+	}
+
+	rmdPodName := ""
+	for _, pod := range pods.Items {
+		rmdPodName = pod.GetObjectMeta().GetName()
+		for _, address := range rmdNode.Status.Addresses {
+			if address.Address == pod.Status.HostIP {
+				break
+			}
+		}
+	}
```
Should also be its own function, same as above... preferably the same function called in both places, but they are in different packages so would need to be placed somewhere that both controllers can import.

nolancon commented on build/manifests/rmd-node-agent-ds.yaml:
```diff
@@ -0,0 +1,38 @@
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+        name: rmd-node-agent-ds
+        namespace: default
+spec:
+        selector:
+                matchLabels:
+                        name: rmd-node-agent-pod
+        template:
+                metadata:
+                        labels:
+                                name: rmd-node-agent-pod
+                spec:
+                        serviceAccountName: "intel-rmd-node-agent"
+                        containers:
+                                - image: intel-rmd-node-agent:latest
+                                  imagePullPolicy: IfNotPresent
+                                  name: rmd-node-agent
+                                  volumeMounts:
+                                  - mountPath: /sys/fs/
+                                    name: cgroup
+                                    mountPropagation: HostToContainer
+                                  env:
+                                  - name: WATCH_NAMESPACE
+                                    value: ""
+                                  - name: POD_NAME
+                                    valueFrom:
+                                            fieldRef:
+                                                    fieldPath: metadata.name
+                                  - name: OPERATOR_NAME
+                                    value: "intel-rmd-node-agent"
+                        volumes:
+                        - name: cgroup
+                          hostPath:
+                                  path: /sys/fs/
+                        nodeSelector:
+                              "feature.node.kubernetes.io/cpu-rdt.RDTL3CA": "true"
```
The formatting is off on this file - large indentations

nolancon reviewed (commented):
Don't worry about getting through all these comments as I know you're pushed for time. They are more for future reference and I will pick up whatever is left in the coming weeks. Great work on this, thanks!

nolancon commented:
Another comment for future reference:
`getPodAddress` function in `rmdworkload_controller` also needs to be updated.

csookarr committed:
Make changes according to code review
Rename rmd-pod-ds.yaml to rmd-ds.yaml

Update funcs that create daemonSets & nodeState to take rmdConfig param

Extra changes to make funcs use rmdConfig as param

Condense call to GET node in nodestate controller

Correct formatting of rmd-node-agent-ds.yaml
